datalad/support/path.py

```diff
@@ -110,4 +113,92 @@ def split_ext(filename):
 
     file_parts = parts[:1] + tail[::-1]
     ext_parts = parts[1+len(tail):]
-    return ".".join(file_parts), "." + ".".join(ext_parts)
\ No newline at end of file
+    return ".".join(file_parts), "." + ".".join(ext_parts)
+
+
+def get_parent_paths(paths, parents, only_with_parents=False):
+    """Given a list of children paths, return their parent paths among parents
+    or their own path if there is no known parent. A path is also considered its
+    own parent (haven't you watched Predestination?) ;)
+
+    All paths should be POSIX, relative, and not pointing outside (not starting
+    with ../)
+
+    Accent is made on performance to avoid O(len(paths) * len(parents))
+    runtime.  ATM should be typically less than O(len(paths) * len(log(parents)))
+
+    Initial intended use - for a list of paths in the repository
+    to provide their paths as files/submodules known to that repository, to
+    overcome difference in ls-tree and ls-files, where ls-files outputs nothing
+    for paths within submodules.
+    It is coded, so it could later be applied even whenever there are nested
+    parents, e.g. parents = ['sub', 'sub/sub'] and then the "deepest" parent
+    is selected
+
+    Parameters
+    ----------
+    parents: list of str
+    paths: list of str
+    only_with_parents: bool, optional
+      If set to True, return a list of only parent paths where that path had
+      a parent
+
+    Returns
+    -------
+    A list of paths (without duplicaates), where some entries replaced with
```

Comment:
@kyleam - if you were to push more, please fix up this typo in `duplicaates`, sorry about that

Reply:
I don't plan on pushing more.  I've got you around the GitPython failures.  Take back your PR :]

Reply:
Great
Then let's take advantage of Germans venturing again to the West and merge it. I will do it locally and fix that typo and then push directlyso we don't stress Travis for no reason